File: DailyChallenges/CPP/LeetCode_45.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) return 0;
        int ans = 1, l = 0, r = nums[0];
        while (r < n-1){
            int _l = l, _r = r; // old value
            while (++_l <= _r){
                if (_l + nums[_l] > r){
                    l = _l;
                    r = l + nums[l];
                }
            }
            ++ans;
        }
        return ans;
    }
};

File: DailyChallenges/Scala/LeetCode_45.scala
object Solution {
    def jump(nums: Array[Int]): Int = {
        val n = nums.length
        def loop(i: Int = 0, jumps: Int = 0, end: Int = 0, farthest: Int = 0): Int = if (i == n-1) jumps else {
            val _farthest = farthest max (i+nums(i)) 
            if (i == end)
                loop(i+1, jumps+1, _farthest, _farthest)
            else
                loop(i+1, jumps, end, _farthest)
        }
        loop()
    }
}
